MetaMorpheus/EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -42,42 +51,99 @@ public static class PEP_Analysis_Cross_Validation
+                IDataView dataView = mlContext.Data.LoadFromEnumerable(PSMDataGroups[allGroupIndexes[0]].Concat(PSMDataGroups[allGroupIndexes[1]].Concat(PSMDataGroups[allGroupIndexes[2]])));
+                trainedModels[groupIndexNumber] = pipeline.Fit(dataView);
+                var myPredictions = trainedModels[groupIndexNumber].Transform(mlContext.Data.LoadFromEnumerable(PSMDataGroups[groupIndexNumber]));
+                CalibratedBinaryClassificationMetrics metrics = mlContext.BinaryClassification.Evaluate(data: myPredictions, labelColumnName: "Label", scoreColumnName: "Score");
+
+                //Parallel operation of the following code requires the method to be stored and then read, once for each thread
+                //if not output directory is specified, the model cannot be stored, and we must force single-threaded operation
+                if (OutputFolder != null)
+                {
+                    mlContext.Model.Save(trainedModels[groupIndexNumber], dataView.Schema, Path.Combine(OutputFolder, "model.zip"));
+                }
+
+                //model is trained on peptides but here we can use that to compute PEP for all PSMs
+                int ambiguousPeptidesResolved = Compute_PSM_PEP(peptideGroups, peptideGroupIndices[groupIndexNumber], mlContext, trainedModels[groupIndexNumber], SearchType, OutputFolder);
+
+                allMetrics.Add(metrics);
+                sumOfAllAmbiguousPeptidesResolved += ambiguousPeptidesResolved;
             }
-            
 
+            return AggregateMetricsForOutput(allMetrics, sumOfAllAmbiguousPeptidesResolved);
+        }
+
+        /// <summary>
+        /// Sets the following static properties: ChargeStateMode, FileSpecificMedianFragmentMassErrors, FileSpecificTimeDependantHydrophobicityAverageAndDeviation_unmodified, FileSpecificTimeDependantHydrophobicityAverageAndDeviation_modified, and FileSpecificTimeDependantHydrophobicityAverageAndDeviation_CZE
+        /// </summary>
+        /// <param name="trainingData"> The PSMs that will be used for training </param>
+        /// <param name="trainingVariables"> An array of training variables from PsmData.trainingInfos dictionary </param>
+        public void BuildFileSpecificDictionaries(List<SpectralMatch> trainingData, string[] trainingVariables)
+        {
+            FileSpecificMedianFragmentMassErrors = GetFileSpecificMedianFragmentMassError(trainingData);
+            ChargeStateMode = GetChargeStateMode(trainingData);
+            
             //These two dictionaries contain the average and standard deviations of hydrophobicitys measured in 1 minute increments accross each raw
             //file separately. An individully measured hydrobophicty calculated for a specific PSM sequence is compared to these values by computing
             //the z-score. That z-score is used as a feature for machine learning.
```

Comment:
Can we move/copy these descriptions to the dictionary declarations at the top of the class?